### ISSUE TITLE
Tests within a Collection don't get same class fixture instance when triggered with Ctrl + R, Ctrl + T

### ISSUE THREAD

**Issue by benmccallum** [xunit/visualstudio.xunit#287]
Sorry if this is an issue in VS, but not sure where it's arising.

I have the following setup, essentially multiple tests in a single test class, and a fixture which seeds a record that can only be seeded once, else the db will complain about uniqueness.

If I run this from the Test Explorer in VS2022, e.g. right-click the class and run tests, the seed only happens once and all is good.

But, if I run this with my cursor inside the `XyzTests` class, using the keyboard shortcut, it seems like the seed is run multiple times as the insert fails with `Cannot insert duplicate key row in object...`.

It's consistently doing this, so seems like there's an issue somewhere. I don't have VS2019 anymore to see if it would happen there, sorry.

```c# 
[CollectionDefinition(nameof(DbFixture))]
public class DbFixtureCollection : ICollectionFixture<DbFixture>
{

}

public class DbFixture : IAsyncLifetime
{
    public async Task InitializeAsync()
    {
          // Seed a user with email address x that has a unique key
    }
}

[Collection(nameof(DbFixture))]
public class XyzTests
{
    public XyzTests(DbFixture dbFixture) { }

    [Fact]
    public void Fact1() { }

    [Fact]
    public void Fact2() { }
}
```

**Comment by johansandblom** [comment]
I just ran into this problem and I do have VS2019 still installed. I can confirm that in VS2019 Ctrl + R, Ctrl + T does NOT suffer from this problem. It looks like VS2022 is executing the tests in parallel (when triggering via Ctrl + R, Ctrl + T) despite the tests being part of the same xUnit collection.

Adding `[assembly: CollectionBehavior(DisableTestParallelization = true)]` will "solve" the issue but at the cost of having the total test run take longer time, aka not a good solution. :)

@benmccallum: Have you found any more information about this issue?

**Comment by benmccallum** [comment]
@johansandblom , good to know I wasn't going crazy :) I haven't looked into it any further to be honest. 

I'd tag someone from the VS team if I knew who works on this part as it could be their end too I guess.

**Comment by bradwilson** [comment]
I'm unable to repro this. This was my attempted repro:

```csharp
[CollectionDefinition(nameof(MyCollection))]
public class MyCollection : ICollectionFixture<MyFixture>
{ }

public class MyFixture
{
    static volatile int staticInstance = 0;

    public int Instance;

    public MyFixture() =>
        Instance = Interlocked.Increment(ref staticInstance);
}

[Collection(nameof(MyCollection))]
public class UnitTest1
{
    MyFixture fixture;

    public UnitTest1(MyFixture fixture)
    {
        this.fixture = fixture;
    }

    [Fact]
    public void Test1()
    {
        Assert.Equal(1, fixture.Instance);
    }
}

[Collection(nameof(MyCollection))]
public class UnitTest2
{
    MyFixture fixture;

    public UnitTest2(MyFixture fixture)
    {
        this.fixture = fixture;
    }

    [Fact]
    public void Test2()
    {
        Assert.Equal(1, fixture.Instance);
    }
}
```

(I tried several things that were simpler; this was the most complex version I could come up with that might be able to repro the problem, but I was unable to.)

Please provide a full repro project and exact repro steps.

**Comment by benmccallum** [comment]
Sorry for not getting to this yet. Will have a go shortly and see if I can narrow it down or if it's no longer an issue.

**Comment by benmccallum** [comment]
I think I was able to repro this with:
1. the following code, 
2. triggering the test run with Ctrl +R, Ctrl + T (non-debug), with my cursor placed inside the UnitTest2 class
3. On VS2022.

For a lot of our integration tests we use this pattern of a Seeder class inside the test class, which itself takes the `MyFixture`. Potentially this is a problematic pattern? 

Problems are:
1. I repro'd this once, then deleted out `UnitTest1` thinking it wasn't needed, then couldn't repro, added it back and still can't repro. Potentially there's a race condition at play.
4. I feel like if we were using the "Seeder" pattern I would've posted that in the original repro code, so I can't remember if it's necessary.


```
namespace Foo
{
    [CollectionDefinition(nameof(MyCollection))]
    public class MyCollection : ICollectionFixture<MyFixture>
    { }

    public class MyFixture
    {
        static volatile int staticInstance = 0;

        public int Instance;

        public MyFixture() =>
            Instance = Interlocked.Increment(ref staticInstance);
    }

    [Collection(nameof(MyCollection))]
    public class UnitTest1
    {
        MyFixture fixture;

        public UnitTest1(MyFixture fixture)
        {
            this.fixture = fixture;
        }

        [Fact]
        public void Test1()
        {
            Assert.Equal(1, fixture.Instance);
        }
    }

    [Collection(nameof(MyCollection))]
    public class UnitTest2 : IClassFixture<UnitTest2.Seeder>
    {
        MyFixture fixture;
        private Seeder seeder;

        public UnitTest2(
            MyFixture fixture,
            Seeder seeder)
        {
            this.fixture = fixture;
            this.seeder = seeder;
        }

        public class Seeder : IAsyncLifetime
        {
            private MyFixture fixture;

            public Seeder(MyFixture fixture)
            {
                this.fixture = fixture;
            }

            public async Task InitializeAsync()
            {
                await Task.FromResult(0);
                fixture.ToString();
            }

            public async Task DisposeAsync()
            {
                await Task.FromResult(0);
            }
        }

        [Fact]
        public void Test2()
        {
            Assert.Equal(1, fixture.Instance);
        }

        [Fact]
        public void Test3()
        {
            Assert.Equal(1, fixture.Instance);
        }
    }
}
```

**Comment by benmccallum** [comment]
Several runs later and I can't get it to fail again 🤦🏻‍♂️ I'd love to hear thoughts on that Seeder pattern though as we might need to change that up. Thanks again, sorry I've been AWOL.

**Comment by benmccallum** [comment]
I'm being bitten by this finally while trying to write integration tests so I'm slowly expanding the setup we have to try and get to the bottom of it but sheesh... attempt number 3, still unsuccessful.

What does seem consistent though is that keyboard shortcut fails, run via Test Explorer works. 

```c#
using System;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Foo;

public class UtilsFixture : IDisposable, IAsyncDisposable
{
    public void Dispose() { }
    public ValueTask DisposeAsync() => default;
}

public class MyGenericFixture<T>
    : IAsyncLifetime
    where T : class
{
    static volatile int staticInstance = 0;

    public int Instance;

    public MyGenericFixture()
    {
        //Instance = Interlocked.Increment(ref staticInstance);
        Debugger.Break();
    }

    public async Task DisposeAsync() => await Task.CompletedTask;
    public async Task InitializeAsync()
    {
        Instance = Interlocked.Increment(ref staticInstance);
        await Task.CompletedTask;
    }
}

public class ProjectA { }

public class ProjectAFixture : MyGenericFixture<ProjectA>
{

}

[CollectionDefinition(nameof(ProjectACollection))]
public class ProjectACollection
    : ICollectionFixture<ProjectAFixture>
    , ICollectionFixture<UtilsFixture>
{

}

[Collection(nameof(ProjectACollection))]
public class UnitTest1 : IAsyncLifetime
{
    private readonly ProjectAFixture _projectAFixture;
    private readonly UtilsFixture _utilsFixture;

    public UnitTest1(
        ProjectAFixture projectAFixture,
        UtilsFixture utilsFixture)
    {
        _projectAFixture = projectAFixture;
        _utilsFixture = utilsFixture;
    }

    public async Task DisposeAsync() => await Task.CompletedTask;
    public async Task InitializeAsync() => await Task.CompletedTask;

    [Fact]
    public void Test1()
    {
        Assert.Equal(1, _projectAFixture.Instance);
    }

    [Fact]
    public void Test2()
    {
        Assert.Equal(1, _projectAFixture.Instance);
    }

    [Fact]
    public void Test3()
    {
        Assert.Equal(1, _projectAFixture.Instance);
    }
}

```

**Comment by benmccallum** [comment]
What's crazy is somehow in trimming back a real tests class we have I've fixed it. Tried putting my cursor above or between test methods in case cursor placement matters, but I doubt it does, just at this point I'm going mad.

I did the following:
1. Rename the class to have the standard "Tests" suffix as I noticed it didn't.
2. Turned some public properties into private as they should've been.
3. Commented out all the [Fact] methods it had in favour of 3 simple versions just doing `await Task.CompletedResult;`.

Reverting 3 it still works. Reverting 2 it still works... Reverting 1 still works...!!!!!

 and now can't repro there...  I did notice it didn't have `Tests` as a suffix on the class name, but surely that's not it.

I literally

**Comment by bradwilson** [comment]
@benmccallum There is a lot of this process that's unfortunately out of our hands.

I'm convinced the problem is that we're being launched multiple times and asked to run things separately, and of course in that scenario all bets are off on getting us to behave in our normal way. I have no idea how VSTest gets itself into that situation, but I could see it being confused about what it thinks it knows pre-compilation vs. what it does post-compilation causing an "information schism" that makes it think it has two (or more) separate workloads to send to us, rather than just one unified list.

FWIW, under Options > Test > General, I always disable "Discover tests in real time" (and recommend everybody does so, because we don't control that discovery process and it's most definitely flawed for anything other than bone-stock xUnit.net with no extensibility points in use). It's possible that this may be contributing to the confusion (and why I can never seem to repro it), but that's just a guess at this point. I also have essentially 0% confidence that they will open up this source-based discovery system to us, despite having promised to do so many years ago (may be going on half a decade now). The worst part is, our discovery APIs support source-based discovery, as this is how we did it way back in v1 when we first shipped support for Resharper. Any API changes we'd need to make to support their model would be something we'd consider, but at this point it appears that the team is content with their bare minimum implementation and not interested in improvements or opening it up to third parties.
